Revamp Featured page in WooCommerce Marketplace

tgglv opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to 11239-gh-Automattic/woocommerce.com.

We want to revamp the Marketplace page according to [Figma designs](seb2o1DExPxoh5d53ftJif-fi-812%3A13394). Together with the new design, the approach of treating data from WooCommerce.com API was changed as well. Instead of the current approach of sections/containers for products and banners, the new approach uses groups of products displayed one after another. 

Changes:
- Refactor current implementation of featured page blocks (the current version works on custom markup functions like "open container", put some product cards, "close container"; the update uses a simplified approach of showing groups of product cards instead)
- Add markup for showing ratings and review count for products
- Reuses current addon's product card for both search results, products by categories, and featured products
- CSS tweaks to make marketplace product card looks like on the design

### How to test the changes in this Pull Request:

#### Preparation
1. To test these changes properly you need an API for testing.
There're 2 options: 
- Local WCCOM (use 11456-gh-Automattic/woocommerce.com for this option) or
- Hardcoded JSON response from [this script](2a667-pb/)

For the 2nd option save the script in `index.php` file and launch PHP-Builtin server with the command like:
```sh
php -S host.docker.internal:9000
```

2. Disable using transient by changing the first line of the `render_featured` method in `includes/admin/class-wc-admin-addons.php` to: `$featured = false;`

3. Change the API URL used in `$raw_featured = wp_safe_remote_get(...` in the same method from `https://woocommerce.test/wp-json/wccom-extensions/2.0/featured` to `http://host.docker.internal:9000` (you may need to change the function used to `wp_remote_get` as well)

#### Testing
1. Open your testing WP site with WooCommerce on this branch
2. Open WP Admin > WooCommerce > Marketplace page, make sure that you are viewing the "Featured" section 
3. You should see something similar to the image below:
![image](https://user-images.githubusercontent.com/15897544/136460277-195c6afc-c655-4973-9adb-d8954264021b.png)

4. Try extensions search (e.g. via typing "memberships"), the page should look as expected


### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enhancement - Revamp the WooCommerce Marketplace page.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

tgglv committed:
Add stars and reviews number

tgglv committed:
Use icon instead of image if possible

tgglv committed:
Add developed-by string

tgglv committed:
Show ratings with the use of stars

tgglv committed:
Remove the current addons implementation

tgglv committed:
Address review comments

tgglv committed:
Refactor product card to use with groups

tgglv committed:
Handle product-like banners

tgglv committed:
Render buttons based banner

tgglv committed:
Output products according to the group's capacity

tgglv committed:
Fix showing rating/reviews number

tgglv committed:
Fix product-related banner's image position

tgglv committed:
Apply product group's capacity

tgglv committed:
Fix icon/image/paragraph of the product card

tgglv committed:
Adjust group's title+description+see more block

tgglv committed:
Use 3-column groups for screens more than 960px

tgglv committed:
Use currency code from WCCOM instead of symbol

tgglv committed:
Revert debug code

andfinally commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -32,8 +32,9 @@ public static function get_featured() {
 				$headers['Authorization'] = 'Bearer ' . $auth['access_token'];
 			}
 
+			// Important: WCCOM Extensions API v2.0 is used.
```
As mentioned in 11257-gh-Automattic/woocommerce.com#issuecomment-942166271, we want to return the country to the endpoint, like we do on other pages.

```suggestion
			$parameter_string = '';
			$country  = WC()->countries->get_base_country();
			if ( ! empty( $country ) ) {
				$parameter_string = '?' . http_build_query( array( 'country' => $country ) );
			}

			// Important: WCCOM Extensions API v2.0 is used.
```

andfinally commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -32,8 +32,9 @@ public static function get_featured() {
+			// Important: WCCOM Extensions API v2.0 is used.
 			$raw_featured = wp_safe_remote_get(
-				'https://woocommerce.com/wp-json/wccom-extensions/1.0/featured',
+				'https://woocommerce.com/wp-json/wccom-extensions/2.0/featured',
```
```suggestion
				'https://woocommerce.com/wp-json/wccom-extensions/2.0/featured' . $parameter_string,
```

tgglv committed:
Test: clean up before running tests

tgglv committed:
Try remove vendor folder from cache

andfinally committed:
Update includes/admin/class-wc-admin-addons.php

Returning country in request to featured 2.0 endpoint.

andfinally committed:
Update includes/admin/class-wc-admin-addons.php

Adding `$parameter_string` to end of URL for featured endpoint.

andfinally committed:
Updated content-hash to try and fix phpunit failure. https://github.com/woocommerce/woocommerce/pull/30840\#issuecomment-942422984.

andfinally committed:
Merge branch 'trunk' into update/marketplace-featured-page

# Conflicts:
#	includes/admin/class-wc-admin-addons.php
#	includes/admin/views/html-admin-page-addons.php

andfinally committed:
Reverting changes to CircleCi config in 618eca7c6ad76a93bb61972c0bdb55eddff03317 and cc0b78f8ecf878ebc74a20a49c5e29045432050e. These were attempts to unstick PHPUnit tests.

andfinally committed:
Added condition to prevent showing `Sorry, could not find anything` message on featured page. This should only be shown when a search term has been submitted.

andfinally committed:
Restored `output_search_promotion_block` and some of the CSS for `.addons-wcs-banner-block`.

andfinally committed:
Restored more styles for `.addons-wcs-banner-block` used for WCPay and WCShip banners.
Restored `install_woocommerce_services_addon` and `install_woocommerce_payments_addon` methods, needed to install plugins from WCPay and WCShip banners.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -21,9 +21,9 @@ class WC_Admin_Addons {
+	 * @return void
 	 */
-	public static function get_featured() {
+	public static function render_featured() {
```
Here and a few other places we're removing a public method.

I realize this is a pretty specialized class and so it _probably_ is not used by 3PDs/our own extensions, but generally speaking we're pretty careful not to remove public methods without at least deprecating them first.

tgglv replied:
Totally agree with the approach of deprecating public functions, @barryhughes.

The only thing the removed functions were used is to show Add-ons on the Marketplace page. The most of these were added in these PRs: 
- https://github.com/woocommerce/woocommerce/pull/11951
- https://github.com/woocommerce/woocommerce/pull/9750
- https://github.com/woocommerce/woocommerce/pull/15598

These PRs are 4-6 years old. Methods are public to be used in HTML template.
It might be weird to show our promoted products from our Marketplace in other 3PD extensions.

andfinally replied:
Hi @barryhughes, thanks, we've now restored and commented the deprecated methods in https://github.com/woocommerce/woocommerce/pull/30900/commits/85fa25ad929ae9c85598c907c263856637cced7d. 

(We haven't restored any of the CSS classes we're deleting – we think CSS is too fragile for 3PDs to have used.)

barryhughes replied:
Thanks! The only one I think we're missing is `get_featured()`.

(It has been renamed to `render_featured()`, but in principle we should keep a stub or alias with the old name, and mark it deprecated.)

> It might be weird to show our promoted products from our Marketplace in other 3PD extensions.

Yep, I figured it was rather unlikely—but sometimes it's hard to predict what is happening out there (that to say, taking a better safe than sorry approach here 🙃).

tgglv replied:
> Thanks! The only one I think we're missing is get_featured().

Thanks for noticing that, @barryhughes! It was returned back in https://github.com/woocommerce/woocommerce/pull/30900/commits/7cbdd71af636cba5e43e75e4e69dea5038e89844.

> Yep, I figured it was rather unlikely—but sometimes it's hard to predict what is happening out there (that to say, taking a better safe than sorry approach here 🙃).

Indeed. I guess we will ask about these methods across Woo to know more about the possible usage of those methods.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -33,8 +33,15 @@ public static function get_featured() {
 			}
 
+			$parameter_string = '';
+			$country  = WC()->countries->get_base_country();
```
Minor PHPCS issue:

```suggestion
			$country          = WC()->countries->get_base_country();
```

tgglv replied:
Thanks, applied in https://github.com/woocommerce/woocommerce/pull/30900/commits/f54d7b0991520208fa826f1147e29d6fa04e27f5.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -127,534 +133,111 @@ public static function get_sections() {
 	 */
-	public static function output_banner_block( $block ) {
+	private static function output_group( $block ) {
+		$capacity           = $block->capacity ?? 3;
```
Minor thing (PHPCS):

```suggestion
		$capacity             = $block->capacity ?? 3;
```

tgglv replied:
Applied in https://github.com/woocommerce/woocommerce/pull/30900/commits/f54d7b0991520208fa826f1147e29d6fa04e27f5, thanks!

barryhughes commented on includes/admin/views/html-admin-page-addons.php:
```diff
@@ -16,6 +16,40 @@
+	if ( $rating >= $index ) {
+		// Rating more that current star to show.
+		return 'fill';
+	} else if (
```
I think this function is being relocated via another PR, but worth flagging this should be `elseif` (no space), because current PHPCS rules.

tgglv replied:
Indeed. Fixed in https://github.com/woocommerce/woocommerce/pull/30900/commits/f54d7b0991520208fa826f1147e29d6fa04e27f5.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -950,4 +510,190 @@ public static function format_promotions( array $promotions ): array {
+									<?php $rating_star_class = 'product-rating-star product-rating-star__' . wccom_get_star_class( $mapped->rating, $index ); ?>
+									<div class="<?php echo esc_attr( $rating_star_class ); ?>"></div>
+								<?php endfor; ?>
+								<span class="product-reviews-count">(<?php echo wp_kses_post( $mapped->reviews_count ); ?>)</span>
```
I don't have a local WPCOM setup to test this against, but I wondered: is `$mapped->reviews_count` an integer?

If it is, could we use a less expensive cast like `(int)` or else the `absint()` function, and/or if it is more complex than that but still does not contain HTML, can we use `esc_html()` (also less expensive)?

andfinally replied:
Thanks, it is an integer – fixed in https://github.com/woocommerce/woocommerce/pull/30900/commits/22b67783d0cecbefb601b7dbf5f0c41cc64c772f.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -950,4 +510,190 @@ public static function format_promotions( array $promotions ): array {
+							<?php if ( $mapped->is_free ) : ?>
+								<span class="price"><?php esc_html_e( 'Free', 'woocommerce' ); ?></span>
+							<?php else : ?>
+								<span class="price"><?php echo wp_kses_post( $mapped->price ); ?></span>
```
Do we need `wp_kses_post()` here or would `esc_html()` suffice?

andfinally replied:
We've addressed this in https://github.com/woocommerce/woocommerce/pull/30900/commits/124dc35e42b93020496f86f7e6dcf679860637c0 – we had to use `wp_kses` instead of `esc_html`, as this value contains `span` and `bdi` tags.

barryhughes commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -950,4 +510,190 @@ public static function format_promotions( array $promotions ): array {
+										sprintf(
+											'<a class="product-vendor-link" href="%1$s" target="_blank">%2$s</a>',
+											esc_url_raw( $mapped->vendor_url ),
+											wp_kses_post( $mapped->vendor_name )
```
Is `esc_html()` sufficient here?

andfinally replied:
Yes thanks, changed in https://github.com/woocommerce/woocommerce/pull/30900/commits/22b67783d0cecbefb601b7dbf5f0c41cc64c772f.

barryhughes reviewed (changes requested):
Looks good! Left a few notes—main concern would be the removal of public methods without a deprecation stage (but, perhaps that is ok for `WC_Admin_Addons`).

tgglv commented:
> > Hardcoded JSON response from [this script](https://github.com/woocommerce/woocommerce/pull/2a667-pb/)
> 
> Hi @tgglv—can you confirm the URL here is correct? It leads me to a 404.

Thanks for the review @barryhughes. 
Don't know why the link was changed. Actually, it's PasteBin: https://mc.a8c.com/pb/2a667/

tgglv committed:
Apply PHPCS suggestions

andfinally committed:
Restored and annotated deprecated functions for backwards compatibility.

andfinally committed:
Addressing feedback. Using cheaper escaping methods.

andfinally committed:
Correcting escaping of `$mapped->price`, using `wp_kses`.

andfinally committed:
Merge branch 'trunk' into update/marketplace-featured-page

andfinally committed:
Added bottom margin of paragraph in `.addons-buttons-banner-details`, to improve rendering of promotional banner on featured page.

andfinally committed:
Merge branch 'update/marketplace-featured-page' of https://github.com/woocommerce/woocommerce into update/marketplace-featured-page

barryhughes reviewed (changes requested):
One last note from me (re `::get_featured()`), plus we need to resolve conflicts per note from @Konamiman.

Those things done, it should be good to merge :+1:

tgglv committed:
Merge remote-tracking branch 'origin/trunk' into update/marketplace-featured-page

tgglv committed:
Fix after merge with trunk

tgglv committed:
Revert get_featured method

barryhughes reviewed (approved):
LGTM!

barryhughes commented:
@andfinally I didn't want to merge without your approval on the PR, but we're all set from my perspective.
